ci/compute-matrix.py

```diff
@@ -0,0 +1,714 @@
+        new_job[explode_tag] = value
+        result.extend(explode_tags(new_job))
+
+    return result
+
+
+def preprocess_matrix_jobs(matrix_jobs):
+    result = []
+    for matrix_job in matrix_jobs:
+        validate_required_tags(matrix_job)
+        set_default_tags(matrix_job)
+        for job in explode_tags(matrix_job):
+            set_derived_tags(job)
+            # The derived tags may need to be exploded again:
+            result.extend(explode_tags(job))
+    return result
+
+
+def filter_projects(matrix_jobs, projects):
+    return [job for job in matrix_jobs if job['project'] in projects]
+
+
+# Skip test jobs that are defined in the matrix.yaml['skip_test_jobs'] list.
+# If a skipped test job is a build_required_job, replace it with the 'build' job.
+# If a skipped test job is not a build_required_job, ignore it.
+# If no jobs remain, remove the matrix_job.
+def skip_test_jobs(matrix_jobs):
+    new_matrix_jobs = []
+    for matrix_job in matrix_jobs:
+        jobs = matrix_job['jobs']
+        new_jobs = set()
+        for job in jobs:
+            if job in matrix_yaml['skip_test_jobs']:
+                if job in matrix_yaml['build_required_jobs']:
+                    # Replace with the prerequisite build job:
+                    new_jobs.add('build')
+                else:
+                    pass  # Ignore the job
+            else:
+                new_jobs.add(job)
+        if new_jobs:
+            new_matrix_job = copy.deepcopy(matrix_job)
+            new_matrix_job['jobs'] = list(new_jobs)
+            new_matrix_jobs.append(new_matrix_job)
+    return new_matrix_jobs
+
+
+def generate_dispatch_group_name(matrix_job):
+    project_name = get_formatted_project_name(matrix_job['project'])
+    ctk = matrix_job['ctk']
+    device_compiler = matrix_job['cudacxx']
+    host_compiler_name = get_formatted_host_compiler_name(matrix_job['cxx'])
+
+    compiler_info = ""
+    if device_compiler['name'] == 'nvcc':
+        compiler_info = f"nvcc {host_compiler_name}"
+    elif device_compiler['name'] == 'llvm':
+        compiler_info = f"clang-cuda"
+    else:
+        compiler_info = f"{device_compiler['name']}-{device_compiler['version']} {host_compiler_name}"
+
+    return f"{project_name} {compiler_info} CTK{ctk}"
+
+
+def generate_dispatch_job_name(matrix_job, job_type):
+    std_str = ("C++" + str(matrix_job['std']) + " ") if 'std' in matrix_job else ''
+    cpu_str = matrix_job['cpu']
+    gpu_str = (', ' + matrix_job['gpu'].upper()) if job_type in matrix_yaml['gpu_required_jobs'] else ""
+    cuda_compile_arch = (" sm{" + matrix_job['sm'] + "}") if 'sm' in matrix_job else ""
+    cmake_options = (' ' + matrix_job['cmake_options']) if 'cmake_options' in matrix_job else ""
+
+    host_compiler_name = get_formatted_host_compiler_name(matrix_job['cxx'])
+    host_compiler_info = f"{host_compiler_name}{matrix_job['cxx']['version']}"
+
+    config_tag = f"{std_str}{host_compiler_info}"
```

Comment:
This will probably trigger a discussion, but having taken a look at the CI results on this PR, I would very much prefer these two to be swapped. At least within libcu++, I am much more concerned about recognizing which _compilers_ as a group are failing, rather than the standard version. When things fail across a language version, I expect the pattern to be immediately visible, whereas for noticing that it's a compiler version problem, if this stays as is, I now need to parse the text in the job titles, and it's not going to be a constant number of lines to jump over to see the pattern, because different compiler versions support different sets of standard versions.

I understand that this may very well be backwards vs what the current experience in the other projects is, so if most folks would prefer it the way it is currently in the PR, that's okay; but if people didn't much care either way, my vote would be to flip the order in this string.

Reply:
I expect some discussion around this and don't have strong opinions myself (other than that I'd *love* for GH to support having multiple nested groupings of job, but alas...)

Let's collect some opinions and hash this out in next week's meeting. It'll be easy to change this at any time.